DATA ANALYSIS:

1. The Given data provide various details (series) like CO2 emissions per capita (metric tons), Energy use per capita (kilograms of oil equivalent), Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $) etc. of differrents countries and each series has it's own series code.

2. The other columns are scale, decimals, country name, and data of years 1990 to 2011.

3. We could also notice some values missing or marked ".." or NaN which need to be cleaned appropiately use python coding.


ANALYSING DATA USING PYTHON:

1. Import necessary modules.
2. Load in your excel datasheet.
3. Check no. of rows and columns.
4. Get the names of columns.
5. Get data type of each column and check if it's according to our requirement.
6. Overview of the data by checking the first ten rows and ensuring suitable data for them, so that they can be further optimized.
7. Descriptive analysis of all the columns.
8. Using Unique() function in each column to check of anamoly/ies.

STEP 1 : Import liberaries and loading up the excel datasheet

In [26]:
import pandas as pd
import numpy as np

#LOADING DATA FROM EXCEL FILE
# define the file name and the data sheet
orig_data_file = r"Original Data.xls"
data_sheet = "Data"

# read the data from the excel file to a pandas DataFrame
data_orig = pd.read_excel(io=orig_data_file, sheet_name=data_sheet)

STEP 2 : GLOBAL DATA OVERVIEW

-> Shape of excel sheet.

-> Availalble columns and their datatypes.

-> Overview of first 10 rows.

-> Descriptive statistics of the excel sheet.

-> Using Unique() to check for irregularities in different columns and thus aknowledging them.

In [ ]:
#-> Shape of excel sheet.
print("Shape of the original dataset:",data_orig.shape)
print("\n")
print("\n")

#-> Availalble columns and their datatypes.
print("Available columns:",data_orig.columns)
print("\n")
print("\n")
print("Column data types:",data_orig.dtypes)
print("\n")
print("\n")

#-> Overview of first 10 rows.
print("Overview of the first 5 rows:")
data_orig.head()
print("\n")
print("\n")

#-> Descriptive statistics of the excel sheet.
print("Descriptive statistics of the columns:")
data_orig.describe()
print("\n")
print("\n")

#-> Using Unique() to check for irregularities in different columns and thus aknowledging them.
data_orig['Series name'].unique()
print("\n")
print("\n")
data_orig['Series code'].unique()
print("\n")
print("\n")
data_orig['SCALE'].unique()
print("\n")
print("\n")
data_orig['Decimals'].unique()
print("\n")
print("\n")

#-> Aknowledging irregularities in the columns.
data_orig[data_orig['SCALE']=='Text']
print("\n")
print("\n")
data_orig[data_orig['Decimals']=='Text']
print("\n")
print("\n")


Shape of the original dataset: (13512, 28)




Available columns: Index(['Country code', 'Country name',  'Series code',  'Series name',
              'SCALE',     'Decimals',           1990,           1991,
                 1992,           1993,           1994,           1995,
                 1996,           1997,           1998,           1999,
                 2000,           2001,           2002,           2003,
                 2004,           2005,           2006,           2007,
                 2008,           2009,           2010,           2011],
      dtype='object')




Column data types: Country code    object
Country name    object
Series code     object
Series name     object
SCALE           object
Decimals        object
1990            object
1991            object
1992            object
1993            object
1994            object
1995            object
1996            object
1997            object
1998            object
1999            object
2000            object


STEP 3 : FINDING FROMM THE GLOBAL OVERVIEW:

-> This global overview gives away the following facts about the available data:
* shape: 28 columns, 13512 rows
* all columns are of type "object" - neither numeric, nor string/text values
* A certain amount of missing values, denoted both as NaN (not a number values) and as the string ".."
* The rows marked as *'Text'* in the columns *'SCALE'* and *'Decimals'* do not contain any information, almost completely composed of NaN values
* The columns represent key values such as country, but also the corresponding years and the series code/name
* The columns *'Country name'*, *'Series code'*, *'SCALE'* and *'Decimals'* do not give any information and are therefore obsolete
* The column *'Series name'* contains the country-specific features required for the analysis
* The names of the features in the column *'Series name'* are clear but too long


DATA CLEANING: 

The objective is to restructure the dataset by turning the 'Series name' values into distinct features, ensuring each row corresponds to a unique combination of country and year. To enhance clarity, year values will be consolidated into a single column.

To ensure data integrity and consistency, the following preprocessing steps will be taken:
- Eliminate any rows where the "SCALE" or "Decimals" columns contain the text "Text".
- Remove extraneous columns: "Country name", "Series code", "SCALE", and "Decimals".
- Replace all instances of ".." and blank cells ("") with NaN to standardize missing data.
- Convert data columns to numerical types for accurate processing.
- Clean and rename entries in the "Series name" column to create valid and interpretable feature names.


STEPS IN DATA CLEANING:

1.  Create a new Data Frame object and alter it by removing entries with "SCALE" and "DECIMAL" values assigned as "TEXT" are deleted. Thus ensure this by checking the no. of rows and columns in original and new file.
2. Drop 'Country name', 'Series code', 'SCALE' and  'Decimals' columns as they are not required for our program.
3. Replace "" and ".." object values with NaN to avoid misinterpretetion by python.
4. Convert columns 1990 to 2011 to numeric data type.
5. Rename the features in column "Series name":

   The variable/feature names in the column *'Series name'* are clear, but too long and not practical to use in the code. In order to improve that, the most relevant feature names will be renamed with shorter labels as indicated in the following table:

<table>
    <tr>
        <td> <b>Variable name</b> </td>
        <td> <b>Description</b> </td>
        <td> <b>Unit</b> </td>
    </tr>
    <tr>
        <td> cereal_yield </td>
        <td> Cereal yield </td>
        <td> kg per hectare </td>
    </tr>
    <tr>
        <td> fdi_perc_gdp </td>
        <td> Foreign direct investment, net inflows </td>
        <td> % of GDP </td>
    </tr>
    <tr>
        <td> elec_access_perc </td>
        <td> Access to electricity </td>
        <td> % of total population </td>
    </tr>
    <tr>
        <td> en_per_gdp </td>
        <td> Energy use per units of GDP </td>
        <td> kg oil eq./\$1,000 of 2005 PPP \$ </td>
    </tr>
    <tr>
        <td> en_per_cap </td>
        <td> Energy use per capita </td>
        <td> kilograms of oil equivalent </td>
    </tr>
    <tr>
        <td> co2_ttl </td>
        <td> CO2 emissions, total </td>
        <td> KtCO2 </td>
    </tr>
    <tr>
        <td> co2_per_cap </td>
        <td> CO2 emissions, total </td>
        <td> metric tons </td>
    </tr>
    <tr>
        <td> co2_per_gdp </td>
        <td> CO2 emissions per units of GDP </td>
        <td> kg/\$1,000 of 2005 PPP \$ </td>
    </tr>
    <tr>
        <td> other_ghg_ttl </td>
        <td> Other GHG emissions, total </td>
        <td> KtCO2e </td>
    </tr>
    <tr>
        <td> ch4_ttl </td>
        <td> Methane (CH4) emissions, total </td>
        <td> KtCO2 </td>
    </tr>
    <tr>
        <td> n2o_ttl </td>
        <td> Nitrous oxide (N2O) emissions, total </td>
        <td> KtCO2 </td>
    </tr>
    <tr>
        <td> nat_emerg </td>
        <td> Droughts, floods, extreme temps </td>
        <td> % pop. avg. 1990-2009 </td>
    </tr>
    <tr>
        <td> pop_urb_aggl_perc </td>
        <td> Population in urban agglomerations >1million </td>
        <td> % </td>
    </tr>
    <tr>
        <td> prot_area_perc </td>
        <td> Nationally terrestrial protected areas </td>
        <td> % of total land area </td>
    </tr>
    <tr>
        <td> gdp </td>
        <td> Gross Domestic Product (GDP) </td>
        <td> \$ </td>
    </tr>
    <tr>
        <td> gni_per_cap </td>
        <td> GNI per capita </td>
        <td> Atlas \$ </td>
    </tr>
    <tr>
        <td> under_5_mort_rate </td>
        <td> Under-five mortality rate </td>
        <td> per 1,000 </td>
    </tr>
    <tr>
        <td> pop_growth_perc </td>
        <td> Population growth </td>
        <td> annual % </td>
    </tr>
    <tr>
        <td> pop </td>
        <td> Population </td>
        <td> 1 </td>
    </tr>
    <tr>
        <td> urb_pop_growth_perc </td>
        <td> Urban population growth </td>
        <td> annual % </td>
    </tr>
    <tr>
        <td> urb_pop </td>
        <td> Urban population </td>
        <td> 1 </td>
    </tr>
    
</table>

In [ ]:
# assign the data to a new DataFrame, which will be modified
data_clean = data_orig

print("Original number of rows:",data_clean.shape[0])
print("\n")
print("\n")

# remove rows characterized as "Text" in the SCALE column
data_clean = data_clean[data_clean['SCALE']!='Text']
data_clean = data_clean[data_clean['Decimals']!='Text']

print("Unnecessary rows removed from the SCALE and DECIMALS column.")
print("Current number of rows:")
print(data_clean.shape[0])
print("\n")
print("\n")

# remove the columns that are not needed for further analysis
# 'Country name', 'Series code', 'SCALE', and 'Decimals' are not needed for the analysis
print("Original number of columns:")
print(data_clean.shape[1])

data_clean = data_clean.drop(['Country name', 'Series code', 'SCALE', 'Decimals'], axis='columns')

print("Current number of columns:")
print(data_clean.shape[1])

# replace empty strings and '..' with NaN in the data columns
data_clean.iloc[:,2:] = data_clean.iloc[:,2:].replace({'':np.nan, '..':np.nan})

# convert the data columns to numeric, forcing errors to NaN
data_clean2 = data_clean.applymap(lambda x: pd.to_numeric(x, errors='ignore'))
# Errors are ignored in order to avoid error messages about the first two columns, which don't need to be transformed
# into numeric type anyway

print("Print the column data types after transformation:","\n",data_clean2.dtypes)
print("\n")
print("\n")

# define shorter names corresponding to most relevant variables in a dictionary
chosen_vars = {'Cereal yield (kg per hectare)': 'cereal_yield',
               'Foreign direct investment, net inflows (% of GDP)': 'fdi_perc_gdp',
               'Access to electricity (% of total population)': 'elec_access_perc',
               'Energy use per units of GDP (kg oil eq./$1,000 of 2005 PPP $)': 'en_per_gdp',
               'Energy use per capita (kilograms of oil equivalent)': 'en_per_cap',
               'CO2 emissions, total (KtCO2)': 'co2_ttl',
               'CO2 emissions per capita (metric tons)': 'co2_per_cap',
               'CO2 emissions per units of GDP (kg/$1,000 of 2005 PPP $)': 'co2_per_gdp',
               'Other GHG emissions, total (KtCO2e)': 'other_ghg_ttl',
               'Methane (CH4) emissions, total (KtCO2e)': 'ch4_ttl',
               'Nitrous oxide (N2O) emissions, total (KtCO2e)': 'n2o_ttl',
               'Droughts, floods, extreme temps (% pop. avg. 1990-2009)': 'nat_emerg',
               'Population in urban agglomerations >1million (%)': 'pop_urb_aggl_perc',
               'Nationally terrestrial protected areas (% of total land area)': 'prot_area_perc',
               'GDP ($)': 'gdp',
               'GNI per capita (Atlas $)': 'gni_per_cap',
               'Under-five mortality rate (per 1,000)': 'under_5_mort_rate',
               'Population growth (annual %)': 'pop_growth_perc',
               'Population': 'pop',
               'Urban population growth (annual %)': 'urb_pop_growth_perc',
               'Urban population': 'urb_pop'
                }

# rename all variables in the column "Series name" with comprehensible shorter versions
data_clean2['Series name'] = data_clean2['Series name'].replace(to_replace=chosen_vars)


Original number of rows: 13512




Unnecessary rows removed from the SCALE and DECIMALS column.
Current number of rows:
10017




Original number of columns:
28
Current number of columns:
24


C:\Users\shiva\AppData\Local\Temp\ipykernel_25668\448712904.py:29: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data_clean.iloc[:,2:] = data_clean.iloc[:,2:].replace({'':np.nan, '..':np.nan})
C:\Users\shiva\AppData\Local\Temp\ipykernel_25668\448712904.py:32: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data_clean2 = data_clean.applymap(lambda x: pd.to_numeric(x, errors='ignore'))
C:\Users\shiva\AppData\Local\Temp\ipykernel_25668\448712904.py:32: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  data_clean2 = data_clean.applymap(lambda x: pd.to_numeric(x, errors='ignore'))


Print the column data types after transformation: Country code     object
Series name      object
1990            float64
1991            float64
1992            float64
1993            float64
1994            float64
1995            float64
1996            float64
1997            float64
1998            float64
1999            float64
2000            float64
2001            float64
2002            float64
2003            float64
2004            float64
2005            float64
2006            float64
2007            float64
2008            float64
2009            float64
2010            float64
2011            float64
dtype: object






DATA CLEANING : STAGE 2 : 

* Analysing the datasheet we observe we can alter it's orientation to improve readablity and ease of apply coding techniques to it. 

In [29]:
# save the short feature names into a list of strings
chosen_cols = list(chosen_vars.values())

# define an empty list, where sub-dataframes for each feature will be saved
frame_list = []

# iterate over all chosen features
for variable in chosen_cols:

    # pick only rows corresponding to the current feature
    frame = data_clean2[data_clean2['Series name'] == variable]

    # melt all the values for all years into one column and rename the columns correspondingly
    frame = frame.melt(id_vars=['Country code', 'Series name']).rename(columns={'Country code': 'country', 'variable': 'year', 'value': variable}).drop(['Series name'], axis='columns')

    # add the melted dataframe for the current feature into the list
    frame_list.append(frame)


# merge all sub-frames into a single dataframe, making an outer binding on the key columns 'country','year'
from functools import reduce
all_vars = reduce(lambda left, right: pd.merge(left, right, on=['country','year'], how='outer'), frame_list)


data_clean2.head()

Country code Country name     Series code  \
0          ABW        Aruba  AG.LND.EL5M.ZS   
1          ADO      Andorra  AG.LND.EL5M.ZS   
2          AFG  Afghanistan  AG.LND.EL5M.ZS   
3          AGO       Angola  AG.LND.EL5M.ZS   
4          ALB      Albania  AG.LND.EL5M.ZS   

                           Series name  SCALE  Decimals       1990  1991  \
0  Land area below 5m (% of land area)      0         1  29.574810   NaN   
1  Land area below 5m (% of land area)      0         1   0.000000   NaN   
2  Land area below 5m (% of land area)      0         1   0.000000   NaN   
3  Land area below 5m (% of land area)      0         1   0.208235   NaN   
4  Land area below 5m (% of land area)      0         1   4.967875   NaN   

   1992  1993  ...  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  
0   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 28 columns]

DATA CLEANING : STAGE 3 : 

* Handling Missing values but also trying to preserve the most of the data.

-> Step 1: find total no. of missing values in each columns and sort them is ascending order.

-> Step 2: Find total no. of missing values for each year rows.

-> Step 3: Analyze the outcomes.

-> Step 4: Optimize the data accordingly.


In [30]:
print("check the amount of missing values in each column")
all_vars.isnull().sum()




check the amount of missing values in each column


country                   0
year                      0
cereal_yield           1377
fdi_perc_gdp           1111
elec_access_perc       5027
en_per_gdp             2082
en_per_cap             1956
co2_ttl                1143
co2_per_cap            1146
co2_per_gdp            1557
other_ghg_ttl          4542
ch4_ttl                4526
n2o_ttl                4526
nat_emerg              4958
pop_urb_aggl_perc      2582
prot_area_perc          726
gdp                     779
gni_per_cap            1013
under_5_mort_rate       716
pop_growth_perc         282
pop                     252
urb_pop_growth_perc     494
urb_pop                 467
dtype: int64

In [31]:
all_vars_clean = all_vars

#define an array with the unique year values
years_count_missing = dict.fromkeys(all_vars_clean['year'].unique(), 0)
for ind, row in all_vars_clean.iterrows():
    years_count_missing[row['year']] += row.isnull().sum()

# sort the years by missing values
years_missing_sorted = dict(sorted(years_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each year
print("missing values by year:")
for key, val in years_missing_sorted.items():
    print(key, ":", val)

missing values by year:
Country name : 2
Decimals : 2
SCALE : 2
Series code : 2
2005 : 1189
2000 : 1273
1995 : 1317
1990 : 1427
2007 : 1631
2006 : 1633
2004 : 1646
2008 : 1708
2003 : 1714
2002 : 1715
2001 : 1718
1999 : 1729
1998 : 1739
1997 : 1746
1996 : 1756
1994 : 1781
1993 : 1792
1992 : 1810
1991 : 1921
2009 : 2078
2010 : 3038
2011 : 4893


DATA OUTCOME ANALYSIS:

* The years 1991 to 2008 offer a good balance and are selected for further processing.

In [36]:
print("number of missing values in the whole dataset before filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the years:")
print(all_vars_clean.shape[0])

print("Column data types:","\n",all_vars_clean.dtypes)
all_vars_clean = all_vars_clean.applymap(lambda x: pd.to_numeric(x, errors='ignore'))
print("Column data types:","\n",all_vars_clean.dtypes)

# filter only rows for years between 1991 and 2008 (having less missing values)
all_vars_clean = all_vars_clean[(all_vars_clean['year'] >= 1991) & (all_vars_clean['year'] <= 2008)]

print("number of missing values in the whole dataset after filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the years:")
print(all_vars_clean.shape[0])

number of missing values in the whole dataset before filtering the years:
41262
number of rows before filtering the years:
6058
Column data types: 
 country                object
year                   object
cereal_yield           object
fdi_perc_gdp           object
elec_access_perc       object
en_per_gdp             object
en_per_cap             object
co2_ttl                object
co2_per_cap            object
co2_per_gdp            object
other_ghg_ttl          object
ch4_ttl                object
n2o_ttl                object
nat_emerg              object
pop_urb_aggl_perc      object
prot_area_perc         object
gdp                    object
gni_per_cap            object
under_5_mort_rate      object
pop_growth_perc        object
pop                    object
urb_pop_growth_perc    object
urb_pop                object
dtype: object


C:\Users\shiva\AppData\Local\Temp\ipykernel_25668\2232159134.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  all_vars_clean = all_vars_clean.applymap(lambda x: pd.to_numeric(x, errors='ignore'))
C:\Users\shiva\AppData\Local\Temp\ipykernel_25668\2232159134.py:7: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  all_vars_clean = all_vars_clean.applymap(lambda x: pd.to_numeric(x, errors='ignore'))


Column data types: 
 country                object
year                   object
cereal_yield           object
fdi_perc_gdp           object
elec_access_perc       object
en_per_gdp             object
en_per_cap             object
co2_ttl                object
co2_per_cap            object
co2_per_gdp            object
other_ghg_ttl          object
ch4_ttl                object
n2o_ttl                object
nat_emerg              object
pop_urb_aggl_perc      object
prot_area_perc         object
gdp                    object
gni_per_cap            object
under_5_mort_rate      object
pop_growth_perc        object
pop                    object
urb_pop_growth_perc    object
urb_pop                object
dtype: object


TypeError: '>=' not supported between instances of 'str' and 'int'

Filtering the countries by missing values:

The same procedure is applied to the filtering of countries with missing values. The following snippet shows the number of NaNs for each country.

In [ ]:
# check the amount of missing values by country

# define an array with the unique country values
countries_count_missing = dict.fromkeys(all_vars_clean['country'].unique(), 0)

# iterate through all rows and count the amount of NaN values for each country
for ind, row in all_vars_clean.iterrows():
    countries_count_missing[row['country']] += row.isnull().sum()

# sort the countries by missing values
countries_missing_sorted = dict(sorted(countries_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each country
print("missing values by country:")
for key, val in countries_missing_sorted.items():
    print(key, ":", val)

missing values by country:
CHL : 120
ISR : 120
AUT : 121
CHE : 121
DEU : 121
DNK : 121
ESP : 121
FIN : 121
FRA : 121
GBR : 121
HUN : 121
IRL : 121
ITA : 121
JPN : 121
KOR : 121
MEX : 121
NLD : 121
PRT : 121
SWE : 121
TUR : 121
USA : 121
AGO : 122
ARG : 122
BGD : 122
BOL : 122
BRA : 122
CAN : 122
CIV : 122
CMR : 122
COG : 122
COL : 122
CRI : 122
DOM : 122
ECU : 122
EGY : 122
EMU : 122
GHA : 122
GRC : 122
GTM : 122
HND : 122
IDN : 122
IND : 122
JOR : 122
KEN : 122
MAR : 122
MOZ : 122
MYS : 122
NGA : 122
PAK : 122
PAN : 122
PER : 122
PHL : 122
PRY : 122
SDN : 122
SEN : 122
SLV : 122
SYR : 122
TGO : 122
THA : 122
TZA : 122
URY : 122
VEN : 122
VNM : 122
ZAF : 122
ZMB : 122
BGR : 123
CHN : 123
LAC : 123
LMC : 123
LMY : 123
MIC : 123
POL : 123
ROM : 123
SAS : 123
SAU : 123
UMC : 123
ZAR : 123
MNA : 124
NZL : 124
AUS : 125
DZA : 125
WLD : 126
ETH : 127
LIC : 127
SSA : 127
YEM : 127
ARE : 131
BEL : 131
ECA : 133
IRN : 133
UKR : 133
LBN : 134
RUS : 134
ARM : 135
BLR : 135
CZE : 135
UZB : 135
KAZ

This output would suggest to remove rows for countries with more than 90 missing values:

In [ ]:
print("number of missing values in the whole dataset before filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the countries:")
print(all_vars_clean.shape[0])


# filter only rows for countries with less than 90 missing values
countries_filter = []
for key, val in countries_missing_sorted.items():
    if val<90:
        countries_filter.append(key)

all_vars_clean = all_vars_clean[all_vars_clean['country'].isin(countries_filter)]

print("number of missing values in the whole dataset after filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the countries:")
print(all_vars_clean.shape[0])

number of missing values in the whole dataset before filtering the countries:
41262
number of rows before filtering the countries:
6058


NameError: name 'countries_missing_sorted' is not defined

DATA CLEANING : STAGE 3 : 
 

* Identifying columns with more no. of NaN values and thus dropping them out.

In [ ]:
# No. of NaN values in each column before filtering
all_vars_clean.isnull().sum()

# remove features with more than 20 missing values

from itertools import compress

# create a boolean mapping of features with more than 20 missing values
vars_bad = all_vars_clean.isnull().sum()>20

# remove the columns corresponding to the mapping of the features with many missing values
all_vars_clean2 = all_vars_clean.drop(compress(data = all_vars_clean.columns, selectors = vars_bad), axis='columns')

print("Remaining missing values per column:")
print(all_vars_clean2.isnull().sum())


# delete rows with any number of missing values
all_vars_clean3 = all_vars_clean2.dropna(axis='rows', how='any')

print("Remaining missing values per column:")
print(all_vars_clean3.isnull().sum())

print("Final shape of the cleaned dataset:")
print(all_vars_clean3.shape)

Remaining missing values per column:
country    0
year       0
dtype: int64


It appears that dropping all features with any no. of missing values doesn't alter the data too much so they are all dropped.

STAGE 4 : EXPORTING THE CLEANED DATASHEET.

In [ ]:
# export the clean dataframe to a csv file
all_vars_clean3.to_csv('data_cleaned.csv', index=False)

NameError: name 'all_vars_clean3' is not defined